Alpha maven and gradle compiler args parsing

MaxCiv opened the pull request.

Commit:
First version

Commit:
ReadMe

Commit:
Migrate to Kotlin

Commit:
Alpha maven and gradle compiler args parsing

Commit:
Gradle parser corrections

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/InferProjectComponent.kt:
```diff
@@ -0,0 +1,24 @@
+ * @since 27.11.2018
+ */
+class InferProjectComponent(project: Project) : ProjectComponent {
+    val pluginSettings: InferPluginSettings? = ServiceManager.getService(project, InferPluginState::class.java).state
```
Мы не хотим тут создать какие-то настройки по умолчанию, чтобы nullable не было?

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/data/report/InferViolation.kt:
```diff
@@ -0,0 +1,35 @@
+class InferViolation {
+
+    @SerializedName("bug_class")
+    var bugClass: String? = null
```
С точки зрения инкапсуляции этот класс --- одна сплошная дыра, с таким количеством публичных изменяемых полей. Нам точно он нужен именно таким?

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/process/InferRunnerImpl.kt:
```diff
@@ -0,0 +1,109 @@
+            BuildTools.GRADLEW -> {
+                javac(filename, GradleParser.updateCompilerArgsForFile(filename, pluginSettings.compilerArgs))
+            }
+            BuildTools.GRADLE -> {
```
Можно в `when` через запятую писать условия, если что =)

MaxCiv replied:
Круто)

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/process/InferRunnerImpl.kt:
```diff
@@ -0,0 +1,109 @@
+            BuildTools.GRADLE -> {
+                javac(filename, GradleParser.updateCompilerArgsForFile(filename, pluginSettings.compilerArgs))
+            }
+            else -> {
```
Зачем?

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/process/InferRunnerImpl.kt:
```diff
@@ -0,0 +1,109 @@
+                gradleClean()
+                gradleRun()
+            }
+            else -> {
```
Зачем?

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/process/InferRunnerImpl.kt:
```diff
@@ -0,0 +1,109 @@
+            else -> {
+            }
+        }
+        pluginSettings.compilerArgs = compilerArgsParser.getCompilerArgs(buildTool, projectPath).toMutableList()
```
А почему тут мы настройки обновляем *после* анализа?

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/process/InferRunnerImpl.kt:
```diff
@@ -0,0 +1,109 @@
+    override fun runAnalysis(buildTool: BuildTools, filename: String): InferReport {
+        when (buildTool) {
+            BuildTools.MAVEN -> {
+                javac(filename, pluginSettings.compilerArgs)
```
В связи с остальными вопросами ниже, можете рассказать, как вообще у вас управляются данные настройки компиляции?

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/process/parsers/CompilerArgsParserImpl.kt:
```diff
@@ -0,0 +1,35 @@
+        val logsFile = File("$projectPath/infer-out/logs")
+        if (!logsFile.exists()) return listOf()
+
+        val logLines = BufferedReader(FileReader(logsFile) as Reader?).readLines()
```
А зачем тут каст в `Reader?`?

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/process/parsers/GradleParser.kt:
```diff
@@ -0,0 +1,54 @@
+        val javacRegex = """.*javac\s*""".toRegex()
+        val javaFilesRegex = """@.*/gradle.*txt""".toRegex()
+        val javacOptionRegex = """(\s|^)-\S+""".toRegex()
+        val spaceAtBeginningRegex = """^\s.+""".toRegex()
```
Регексы лучше бахнуть приватными полями, чтобы они считались один раз

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/process/parsers/GradleParser.kt:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    fun updateCompilerArgsForFile(filename: String, compilerArgs: List<String>): List<String> {
+        val sourcepathRegex = """.*/main/java""".toRegex()
```
А если внезапно исходники лежат где-то в другом месте?

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/process/parsers/GradleParser.kt:
```diff
@@ -0,0 +1,54 @@
+
+    fun updateCompilerArgsForFile(filename: String, compilerArgs: List<String>): List<String> {
+        val sourcepathRegex = """.*/main/java""".toRegex()
+        val sourcepath = sourcepathRegex.find(filename)!!.value
```
Воу-воу, давайте все же без `!!` =)

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/process/parsers/GradleParser.kt:
```diff
@@ -0,0 +1,54 @@
+                javacOptionRegex.findAll(fullLine)
+                    .forEach { changedLine = changedLine.replace(it.value, "$delimiter${it.value}$delimiter") }
+                changedLine.replace("^\uD83D\uDE31".toRegex(), "")
+                    .replace("$delimiter$".toRegex(), "")
```
Больше регексов в отдельные переменные богу регексов и отдельных переменных

ice-phoenix replied:
Что такое `$delimiter$`?

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/process/parsers/GradleParser.kt:
```diff
@@ -0,0 +1,54 @@
+        return logLines.asSequence()
+            .map { it.replace(loggerMarkerRegex, "") }
+            .dropWhile { !it.contains(javacRegex) }
+            .takeWhile { !it.contains(javaFilesRegex) }
```
А что мы тут делаем?

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/process/parsers/GradleParser.kt:
```diff
@@ -0,0 +1,54 @@
+                    .split(delimiter).asSequence()
+            }
+            .filter { it.isNotEmpty() }
+            .map { if (it.contains(spaceAtBeginningRegex)) it.drop(1) else it }
```
А почему дропаем только один пробел из потенциально многих?

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/process/parsers/GradleParser.kt:
```diff
@@ -0,0 +1,54 @@
+            .flatMap { fullLine ->
+                var changedLine = fullLine
+                javacOptionRegex.findAll(fullLine)
+                    .forEach { changedLine = changedLine.replace(it.value, "$delimiter${it.value}$delimiter") }
```
Что мы тут пытаемся сделать?

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/process/parsers/MavenParser.kt:
```diff
@@ -0,0 +1,28 @@
+        val argsBeginningRegex = """.*\+\+Contents of .*javac\.JavacCompiler\d*arguments.*""".toRegex()
+        val lastPathSeparatorRegex = """([:;])"""".toRegex()
+        val oneArgRegex = """".*?"""".toRegex()
+        val javaFilesRegex = """\.java"$""".toRegex()
```
Регексы лучше считать один раз

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/process/parsers/MavenParser.kt:
```diff
@@ -0,0 +1,28 @@
+            .flatMap {
+                oneArgRegex.findAll(it)
+                    .map { it.value }
+            }
```
Что мы тут делаем?

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/ui/toolwindow/ResultsTab.kt:
```diff
@@ -0,0 +1,65 @@
+class ResultsTab(project: Project) : JPanel(BorderLayout()) {
+    private val pluginSettings: InferPluginSettings? = project.getComponent(InferProjectComponent::class.java).pluginSettings
+    private val treeResults: Tree
+    private var rootNode: RootNode? = null
```
Для подобных случаев есть `lateinit`

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/ui/toolwindow/ResultsTab.kt:
```diff
@@ -0,0 +1,65 @@
+ * @since 01.12.2018
+ */
+class ResultsTab(project: Project) : JPanel(BorderLayout()) {
+    private val pluginSettings: InferPluginSettings? = project.getComponent(InferProjectComponent::class.java).pluginSettings
```
Нам важно, что настройки могут быть `null`?

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/ui/toolwindow/ResultsTab.kt:
```diff
@@ -0,0 +1,65 @@
+        val treeModel = DefaultTreeModel(rootNode)
+        newTree.model = treeModel
+        newTree.cellRenderer = CellRenderer()
+        return newTree
```
Раз дерево так связано с `rootNode`, нам точно нужно держать отдельную ссылку на него? Может быть, `rootNode` можно достать из `tree`?

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/ui/tree/RootNode.kt:
```diff
@@ -0,0 +1,35 @@
+
+    constructor(nodeName: String) : super(nodeName) {}
+
+    internal constructor(inferReport: InferReport) {
```
А почему это не primary конструктор?

ice-phoenix commented on src/main/kotlin/com/maxciv/infer/plugin/ui/tree/RootNode.kt:
```diff
@@ -0,0 +1,35 @@
+    override fun render(cellRenderer: CellRenderer) {
+        cellRenderer.append(LABEL, SimpleTextAttributes.REGULAR_ATTRIBUTES)
+        if (inferReport != null) {
+            cellRenderer.append(" (" + inferReport!!.violations.size.toString() + " violations)",
```
Вы из Java этот код что ли конвертировали? Тут точно `inferReport!!` орет предупреждением, что `!!` не нужен (смарт касты, вся фигня)

ice-phoenix reviewed (commented):
Из основного, что интересно, --- как именно (по сути) вы парсите логи?

Commit:
WIP: parse files and compiler args for modules

Commit:
WIP: working with modules

Commit:
Work with modules

Commit:
Refactoring

Commit:
Refactoring of UI

Commit:
Some actions

Commit:
Class-files analysis

Commit:
Actions, blocking analysis, fixes

Commit:
OnSave analysis and violation highlighting

Commit:
Indicator text, compile on module analysis

Commit:
Icons, result tabs, update report for module

Commit:
All modules analysis

Commit:
--no-progress-bar to infer arguments

Commit:
Multi-files OnSave analysis

Commit:
Magnificent update: choosing infer working dir, different infer w.dir for modules, OnSave analysis with waiting

Next level of incremental analysis, independent infer files for each module, advanced modules analysis

Commit:
Full-project-analysis to Pre-analysis – after capture run all-modules-analysis

Commit:
Compile only current module on Module analysis

Commit:
Short class names option

Commit:
Import/export report

Commit:
Downloading and installing Infer releases from GitHub

Commit:
Gradle plugin for building fat jar

Commit:
Some Kotlin-style perversions

Commit:
Maven/Gradle args, NuProc, OnSaveService, mvnw, mvn install, toggle different resultsDirs

Commit:
Fix

Commit:
Gradle parser fix, configurable capture task

Commit:
Gradle parser and settings fix, get module real name

Commit:
Renaming

Commit:
Merge branch 'develop'
